EasyCommands.Tests/ScriptTests/ScriptTest.cs

```diff
@@ -0,0 +1,182 @@
+        /// </summary>
+        /// <param name="state">The desired end-state.</param>
+        /// <param name="runLimit">The limited number of runs. Defaults to 100</param>
+        public void RunUntil(ProgramState state, int runLimit = 100)
+        {
+            int runCounter = 0;
+            do
+            {
+                MDKFactory.Run(program);
+                runCounter++;
+            } while (Program.STATE != state && runCounter < runLimit);
+        }
+
+        /// <summary>
+        /// Given a custom name for a block (or set of blocks), set up the custom names
+        /// for the given mocked blocks and store them for later retrieval when requested by
+        /// the grid terminal system.
+        /// 
+        /// Since the GridTerminalSystem's GetBlocksOfType method doesn't return values but
+        /// instead modifies the given list, this uses Moq's Callback functionality to
+        /// invoke a callback that will attempt to retrieve the stored blocks.
+        /// </summary>
+        /// <typeparam name="T">The type of the blocks we're mocking retrieval for.</typeparam>
+        /// <param name="name">The custom name of the block(s) we're mocking.</param>
+        /// <param name="blockMocks">Blocked mocks to be stored and used later.</param>
+        public void MockBlocksOfType<T>(String name, params Mock<T>[] blockMocks) where T : class, IMyTerminalBlock
+        {
+            SetupMockBlocksByType(name, blockMocks);
+
+            // Set up the mocked GridTerminalSystem to invoke the appropriate callback method for the appropriate type
+            gridMock.Setup(g => g.GetBlocksOfType(It.IsAny<List<T>>(), It.IsAny<Func<T, bool>>()))
+                .Callback((Action<List<T>, Func<T, bool>>)GetBlocksOfTypeCallback);
+        }
+
+        /// <summary>
+        /// Given a group name, set up a mock group that will be retrieved by the GridTerminalSystem.
+        /// Store the given mocked blocks for later retrieval.
+        /// 
+        /// Also, set up the group's GetBlocksOfType method to use a Moq Callback so that
+        /// those mocked blocks can be found later.
+        /// </summary>
+        /// <typeparam name="T">The type of the blocks being mocked for this group.</typeparam>
+        /// <param name="groupName">The name of the group being mocked.</param>
+        /// <param name="blockMocks">The blocks being mocked and that will be returned later.</param>
+        public void MockBlocksInGroup<T>(String groupName, params Mock<T>[] blockMocks) where T : class, IMyTerminalBlock
+        {
+            var mockGroup = new Mock<IMyBlockGroup>();
+            mockGroup.Setup(g => g.Name).Returns(groupName);
+            mockGroup.Setup(g => g.GetBlocksOfType(It.IsAny<List<T>>(), It.IsAny<Func<T, bool>>()))
+                .Callback((Action<List<T>, Func<T, bool>>)GetBlocksOfTypeCallback);
+            mockedGroups.Add(mockGroup.Object);
+
+            SetupMockBlocksByType(blockMocks);
```

Comment:
I'm trying to think how this will work if I have a script where I want to get the same mocked blocks by name, and by being in group.  I think right now this would double-add to mockedBlocksByName.  

If we change GenericListDictionary to `Dictionary<Type,Set<object.>>` does that alleviate?

Alternatively we could separate adding mock blocks and then adding those blocks to groups, but i like the one shot method.  Thoughts?

Reply:
This is a great catch. Yup! We would double-add.

Changing it to a Set should do the trick here. I've also added a test explicitly around manipulating blocks by-themselves and as a group. Check it out!